### DISCUSSION TITLE
Feature Requests: Inverse Problem Solvers

### DISCUSSION THREAD

**Discussion by kecnry** [phoebe-project/phoebe2#407] [Ideas]
Please include any suggestions for extensions to the inverse problem solvers in PHOEBE here, including:
* new algorithms or codes to include
* new capabilities or options to existing algorithms
* new tutorials or example scripts

Bug reports to existing capabilities belong as a [new issue](https://github.com/phoebe-project/phoebe2/issues/new) instead.  For questions or discussion about existing features, please start a [new discussion](https://github.com/phoebe-project/phoebe2/discussions/new).

For an overview of the existing capabilities, see:
* [PHOEBE 2.3 release paper](http://phoebe-project.org/publications/2020Conroy+)
* [Tutorial: Inverse Problem Overview](http://phoebe-project.org/docs/2.3/tutorials/solver) (and the advanced tutorials linked from the bottom of that page)
* [Tutorial: Distributions](http://phoebe-project.org/docs/2.3/tutorials/distributions)
* [Tutorial: Gaussian Processes](http://phoebe-project.org/docs/2.3/examples/minimal_GPs)

Each individual suggestion/feature should be its own top-level comment so that they can easily be tracked and "voted" to see which have the most interest.  Any conversation about that feature or progress with implementation and/or testing should be as replies to those threads.

**Comment by kecnry** [comment]
Allow for a callable `custom_lnpriors_callable` in addition to the existing `custom_lnprobability_callable` when calling [b.run_solver](http://phoebe-project.org/docs/2.3/api/phoebe.frontend.bundle.Bundle.run_solver).  Although this can currently be handled within `custom_lnprobability_callable`, the current implementation does not allow rejecting a sample *before* running the forward model, which could be useful in some cases.

**Comment by kecnry** [comment]
Convergence stopping criteria for [emcee solver](http://phoebe-project.org/docs/2.3/api/phoebe.parameters.solver.sampler.emcee).  Currently the chains and per-chain autocorrelation times are exposed to the user and it is completely up to them to decide how to determine if the chains have converged sufficiently (and what values of `thin`, `burnin`, and `lnprob_cutoff` are appropriate).

This [emcee tutorial](https://emcee.readthedocs.io/en/stable/tutorials/autocorr/) provides a nice overview of some of the options we may want to consider.

**Reply by kecnry** [reply]
plotting the autocorrelation function has been implemented in 1d92e69b892946f05a6e372c6a672160c55fa626 (currently in the development branch) and should be included in the 2.4 release.

* implements `phoebe.helpers.process_acf`, `process_acf_from_solution`, `acf`, `acf_ci`, `acf_ci_bartlett`
* implements `style='acf'` and `'acf_lnprobabilities'` when plotting emcee solutions

**Comment by kecnry** [comment]
Samplers: ability to expose "warnings" during sampling.  Currently these are only available in the log if the appropriate logger has been initialized.  Options would be to:
* expose in a corner plot similar to `style='failed'` (either in separate corner plot or lumped in with errors but with a different marker?)
* ability to optionally treat warnings as errors and reject them during sampling (would this be across the board or have the ability to choose which warnings would be rejected).  As is, this technically could be done manually through either `custom_lnprobability_callable` (or `custom_lnpriors_callable` (see https://github.com/phoebe-project/phoebe2/discussions/407#discussioncomment-335283) by calling [b.run_checks](http://phoebe-project.org/docs/2.3/api/phoebe.frontend.bundle.Bundle.run_checks) within the custom function... but that would be quite advanced and not exactly convenient.

**Comment by kecnry** [comment]
New Sampler: importance sampling (see [snowline](https://johannesbuchner.github.io/snowline/index.html), for example)

**Comment by kecnry** [comment]
New Samplers: other mcmc implementations besides emcee ([pymc3](https://docs.pymc.io/), [autoemcee](https://johannesbuchner.github.io/autoemcee/index.html))

**Comment by kecnry** [comment]
MCMC automated "branch finder" - to automatically find and isolate the best branch from an MCMC run in use with [resampling](http://phoebe-project.org/docs/2.3/tutorials/emcee_resampling)

**Comment by kecnry** [comment]
Automated backend transitioning - so a solver could occasionally test the residuals between two different backends and transition to more expensive models as needed.  This would need some serious testing and probably would not allow adding new parameters (degrees of freedom) within a single run.

**Comment by kecnry** [comment]
Detrending or data-preprocessing plugins.  In theory, [GPs](http://phoebe-project.org/docs/2.3/examples/minimal_GPs) "should" be used to create a noise-model, but this isn't always practical and noisy data can cause issues with estimators (where you don't have a decent physical model yet).

**Comment by kecnry** [comment]
Bootstrapping algorithms as an alternative method (to mcmc or nested sampling "posteriors") to determine uncertainties

**Reply by jsinkbaek** [reply]
I realize that this is an old comment, but wanted to add a take.

With the current state of computation speed with PHOEBE, these will be expensive since the optimizers usually require a lot of evaluations still.

They have potential use for testing the assumption that the data only includes white noise. If a residual block bootstrap is implemented, the residuals can be shuffled in time/phase in groups to test whether they are biasing the fit.

**Comment by kecnry** [comment]
Main sequence "constraint" (possibly via a "prior" on the relationship between fundamental stellar parameters rather than an exact constraint between parameters).

**Comment by kecnry** [comment]
Additional "observables" or "geometric properties" which could then be included in the likelihood (either by placing priors or by allowing the observed values with uncertainties to be in the cost-function):
* eclipse width/depth/separation (also useful to put a prior on requiring or forbidding presence of eclipse)
* light-ratio (time-dependent or marginalized over orbit?  Or just addressed with per-component fluxes exposed in the light curve?)
* SED or colors (time-dependent or marginalized over orbit?)
* flux-weighted loggs (time-dependent or marginalized over orbit?)
* observed temperatures (time-dependent or marginalized over orbit?) - including contributions from spots, distortion, irradiation, etc.

Many of these could, in theory, currently be handled manually with `custom_lnprobability_callable`.

**Reply by kecnry** [reply]
See #571 and #662 for use-cases where this is also needed to access coupled pblum values.

**Reply by kecnry** [reply]
See also #919 for a case of accessing "observed" temperatures.

**Comment by kecnry** [comment]
GPs: test or prevent GPs from "stealing" signal from the binary model.

**Comment by aprsa** [comment]
Differential corrections. While many scorn at DC, it is demonstrably the fastest way to get to _a_ minimum provided the starting point is reasonably close. DC should be straight-forward to implement and, when used correctly, it would add value to the fitting framework.

**Reply by kecnry** [reply]
DC was implemented and included in the 2.4 release.

**Comment by kecnry** [comment]
When plotting MCMC trace plots, allow coloring by lnprobabilities.  When plotting lnprobabilities vs iteration, allow coloring by any single parameter value.

**Reply by kecnry** [reply]
This has been implemented in f313775e8de28b28693bfd89fb7db2f527bff106 (released in phoebe 2.4).

* if `style='lnprobability'`: `c` can be a twig pointing to any of the sampled parameters (whether within adopt_parameters or not)
* if `style='trace'`: `c` can be a twig or `'lnprobability'`
* NOTE: adding color to lineplots with many walkers/iterations can become expensive

**Comment by kecnry** [comment]
emcee: allow setting init_from to an existing emcee solution instead of requiring calling adopt_solution as an intermediate step.  See the [resampling tutorial](http://phoebe-project.org/docs/2.3/tutorials/emcee_resampling) for the current requirement to call [adopt_solution](http://phoebe-project.org/docs/2.3/api/phoebe.frontend.bundle.Bundle.adopt_solution) to create a tagged distribution first.

**Comment by kecnry** [comment]
fake `progress_every_niters` support in optimizers by (internally) setting a lower `maxiter`, saving the results, and then restarting at the final position until reaching the full provided `maxiter` or canceled.  This would then allow monitoring the progress and would be useful in cases where the desired `xatol` for `fatol` (in the case of nelder-mead) is hard to estimate.. but would likely come with some overhead.

**Reply by kecnry** [reply]
This has now been implemented in aa40e847d2a5309774faac4c304b347a4650f (released in phoebe 2.4).  Restarting wasn't needed - instead the current position of the optimizer is accessed and exported.  The only overhead (besides I/O) are additional calls to compute the current `lnprobability` if `expose_lnprobabilities == True` at each export.

**Comment by kecnry** [comment]
New Samplers: [UltraNest](https://johannesbuchner.github.io/UltraNest/) nested sampling (see #414)

**Comment by JunHuiLiu2017** [comment]


**Reply by kecnry** [reply]
Hi @JunHuiLiu2017 - Are you trying to run the [emcee tutorial](http://phoebe-project.org/docs/2.3/tutorials/emcee) and getting this error message?  The tutorial runs fine for me without any errors.  If you've made any changes from the tutorial, can you send me a script of what you ran that can reproduce this error along with the version of phoebe you're running?  And if you wouldn't mind, can you start a [new issue](https://github.com/phoebe-project/phoebe2/issues/new) so we don't spam this feature request discussion board?  Thanks!

**Comment by kecnry** [comment]
Optimizers: expose "uncertainty estimates" to aid in creating initializing distribution widths for samplers

**Comment by kecnry** [comment]
New optimizers: lmfit, scipy.optimize.least_squares

**Comment by kecnry** [comment]
New estimator: phase-dispersion minimization

**Comment by jsinkbaek** [comment]
It would be interesting to include a bayesian optimization or variational inference algorithm, as a trade-off between wanting a global optimizer (with minimal function evaluations), and approximating the probability distribution.

**Comment by gokhannY** [comment]
I don't know that this is already possible or not, but it would be nice to have light contributions for each component based on phase for each light curve dataset, as in Phoebe 1.0 code. I'm dealing with a lot of spectral disentangling and this feature is a must for me. 

I believe the discussion on a similar topic had been talked in here, but I couldn't find it.

**Reply by kecnry** [reply]
Adding official support for per-component light curve datasets has been on the list for a while, but unfortunately would take a significant amount of work and the concern is that it would further raise the learning curve and make the simpler (more common) case more confusing.  That said, there are some work arounds discussed in #839

**Reply by gokhannY** [reply]
Kyle, I believe the workaround is not the same as requested in #839.

I believe Jeppe was asking the same thing I asked, which getting light contribution of each component in a binary dependent on the phase. Such as WD code gives,
![light_contributions](https://github.com/user-attachments/assets/d2bdf788-220d-4b40-8858-9487e4241163)

**Reply by kecnry** [reply]
You should be able to manually build that same output using the mesh approach explained at the end, but it will be significantly more expensive to compute than a "normal" light curve.

**Reply by gokhannY** [reply]
Hey Kyle,

I tried the mesh approach, but something is off. Could you be able to check the code and reproduce my error?

[my_results](https://github.com/user-attachments/files/18638037/light_contr.pdf)

```import phoebe
import numpy as np
import matplotlib.pyplot as plt

times = np.linspace(0,1,101)

b = phoebe.default_binary()
b.add_constraint('semidetached', 'secondary')

b.flip_constraint('asini@binary', solve_for='sma@binary')
b.add_dataset('lc', compute_times=times, dataset='lc01')
b.add_dataset('mesh', include_times='lc01', columns=['areas', 'mus', 'visibilities', 'intensities@lc01', 'ptfarea@lc01'])

b.set_value('period', component='binary', value=2.043880)
b.set_value('teff', component='primary', value=32000)
b.set_value('teff', component='secondary', value=22300)
b.set_value('requiv', component='primary', value=4.626)
b.set_value_all('incl@binary', 69.851)
b.set_value_all('q', 0.4722)
b.set_value_all('asini@binary', 15.949)

b.set_value_all('atm', 'blackbody')
b.set_value_all('ld_mode', 'manual')
b.set_value_all('ld_mode_bol', 'manual')
b.set_value_all('ld_coeffs', component='primary', value=[0.388, 0.232])
b.set_value_all('ld_coeffs', component='secondary', value=[0.4588, 0.256])

b.set_value_all('pblum_mode', 'component-coupled')
b.set_value('pblum', component='primary', dataset='lc01', value=7.072)

b.run_compute()

lcs = {'primary': np.zeros_like(times), 'secondary': np.zeros_like(times)}
for i, time in enumerate(times):
    for component in ['primary', 'secondary']:
        mesh_ps = b.filter(time=time, component=component, context='model')
        intensities = mesh_ps.get_value('intensities@lc01')
        areas = mesh_ps.get_value('areas')
        mus = mesh_ps.get_value('mus')
        visibilities = mesh_ps.get_value('visibilities')
        ptfarea = mesh_ps.get_value('ptfarea@lc01')
    
        flux = np.nansum(intensities * areas * mus * visibilities) * ptfarea
        lcs[component][i] = flux

plt.plot(times, lcs['primary'], 'b.-')
plt.plot(times, lcs['secondary'], 'r.-')
plt.show()```

**Reply by kecnry** [reply]
what isn't as you'd expect?  You can also plot the sum and see if that agrees with the original light curve (and you may also want to extend to a full cycle, by setting the max on times to be the same as the orbital period).

**Reply by gokhannY** [reply]
I would expect they would be symmetrical at 0 and 1. Like you said in your answer. https://github.com/phoebe-project/phoebe2/discussions/839#discussioncomment-8954138

Obviously, since the temperature and radius of the components are different. But I would expect a similar flux vs phase. 

And also, you said in your comment that it would take more time to compute but it was really quick even though I used more datapoints as you can see from the code. Have I done something wrong?

**Reply by kecnry** [reply]
Can you try with a full orbital cycle and try also plotting the sum compared to the original light curve?

**Reply by gokhannY** [reply]
Happy ending. Thanks for your help Kyle. 
[light_contr.pdf](https://github.com/user-attachments/files/18643381/light_contr.pdf)

**Comment by kecnry** [comment]
Allow for using multiple compute options within a single solver run.  See #877 and #1019 for requested use-cases.
